libpyvinyl/Parameters/Collections.py

```diff
@@ -0,0 +1,318 @@
+        """
+        Create MasterParameter with uninitialized links
+        """
+        self.links = None
+        super().__init__(*args, **kwargs)
+
+    def add_links(self, links):
+        """
+        Links is a dict with key being reference to calculator and name of parameter to overwrite
+        """
+        self.links = links
+
+
+class MasterParameters(Parameters):
+    """
+    Master Parameters object that contain all master parameters with the
+    additional ability to set values for the other parameters this master
+    should control.
+    """
+    def __init__(self, parameters_dict, *args, **kwargs):
+        """
+        Create MasterParameters object with given parameters dict
+
+        The parameters_dict contains all the Parameters objects in the
+        ParametersCollection, and provides the access for the master parameters
+        so they can control the other Parameters objects of which they are
+        responsible.
+        """
+        self.parameters_dict = parameters_dict
+        super().__init__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        """
+        Set item that propagates change throughout all links
+        """
+        master_parameter = self.parameters[key]
+
+        if master_parameter.links is not None:
+            for calculator in master_parameter.links:
+                calculator_par_name = master_parameter.links[calculator]
+                self.parameters_dict[calculator][calculator_par_name] = value
+
+        self.parameters[key].set_value(value)
+
+
+class ParametersCollection(AbstractBaseClass):
+    """
+    Object intended for use as instrument parameters
+
+    This object holds Parameters objects for a number of calculators and can
+    have master parameters which control parameters for a number of
+    calculators at once.
+    """
+    def __init__(self):
+        """
+        Create an empty ParametersCollection instance
+        """
+        self.parameters_dict = {}
+        self.master = MasterParameters(self.parameters_dict)
+
+    @classmethod
+    def from_json(cls, fname: str):
+        """
+        Initialize an instance from a json file.
+
+        :param fname: The filename (path) of the json file.
+        :type  fname: str
+
+        """
+        with open(fname, 'r') as fp:
+            instance = cls.from_dict(json.load(fp))
+
+        return instance
+
+    @classmethod
+    def from_dict(cls, instrument_dict: dict):
+        """
+        Initialize an instance from a dict.
+
+        :param fname: The filename (path) of the json file.
+        :type  fname: str
+
+        """
+        parameters = cls()
+        for key in instrument_dict:
+            parameters.add(key, Parameters.from_dict(instrument_dict[key]))
+
+        return parameters
+
+    def to_dict(self):
+        params_collect = {}
+        for key in self.parameters_dict:
+            params_collect[key] = self.parameters_dict[key].to_dict()
+        return params_collect
```

Comment:
Shouldn't this method store all aspects of the class? It doesn't collect the self.master attribute that contain all the master parameters?

Reply:
we agreed that what matters so far are only the values. But during the week I was trying to see further usage of this class and I might have second thoughts about it. One thing we have not discussed is if we think there might be a benefit from using this classes somewhere else. I suggest to discuss this point during next meeting

Reply:
If a class has a to_dict / from_dict method I would strongly expect it to work for all contents of the class. I don't see why we would limit it to only parts, even if that is all we technically need. Would it take a lot of additional effort to include the full class?

Reply:
It's done in this way.
https://github.com/PaNOSC-ViNYL/libpyvinyl/blob/a3a2599629712b7f6447913911733c8b50edd32b/libpyvinyl/Parameters/Collections.py#L251-L256

https://github.com/PaNOSC-ViNYL/libpyvinyl/blob/a3a2599629712b7f6447913911733c8b50edd32b/libpyvinyl/Parameters/Collections.py#L233-L249

The output .json [file](https://github.com/PaNOSC-ViNYL/libpyvinyl/blob/parameters/notebooks/instrument.json) is quite decent in my view. what do you think?

Reply:
I was just wondering why the value is `null`, isn't it asked at construction?
I think that this might be very dangerous and we should make sure that parameters have always a value assigned.